### PR TITLE
[Doc] Add Datasets Doc

### PR DESCRIPTION
**Pull request by MengzhangLI**

### CODE REVIEW

File: docs/en/advanced_guides/datasets.md [hunk 1]
```diff
@@ -1 +1,107 @@
-# Datasets
+# Dataset
+
+In this document, we will introduce the design of each dataset in MMSegmentation and how users can design their own dataset.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:3]:
```suggestion
In this document, we will introduce the design of datasets in MMSegmentation.
```

File: docs/en/advanced_guides/datasets.md [hunk 2]
```diff
@@ -1 +1,107 @@
...
+
+In the 1.x version of MMSegmentation, all datasets are inherited from `BaseDataset`.
+Each dataset loads the list of data info (e.g., datapath) by `load_data_list`.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:6]:
```suggestion
The function of dataset is loading the `data_list` (please refer ... for more information about `data_list`).
```

File: docs/en/advanced_guides/datasets.md [hunk 3]
```diff
@@ -1 +1,107 @@
-# Datasets
+# Dataset
+
+In this document, we will introduce the design of dataset in MMSegmentation and how users can design their own dataset.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:3]:
```suggestion
In this document, we will introduce the design of dataset in MMSegmentation.
```

File: docs/en/advanced_guides/datasets.md [hunk 4]
```diff
@@ -1 +1,143 @@
-# Datasets
+# Dataset
+
+In this document, we will introduce functions of `BaseSegDataset` class in MMSegmentation.
+
+## BaseSegDataset
+
+`BaseSegDataset` is a custom dataset class for semantic segmentation task, which is inherited from `BaseDataset` in MMEngine.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:7]:
```suggestion
`BaseSegDataset` is a base dataset class for semantic segmentation task, which is inherited from `BaseDataset` in MMEngine.
```

File: docs/en/advanced_guides/datasets.md [hunk 5]
```diff
@@ -1 +1,143 @@
...
+More details of `BaseDataset` could be found in [MMEngine BaseDataset Documentation](https://github.com/open-mmlab/mmengine/blob/main/docs/zh_cn/advanced_tutorials/basedataset.md).
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:8]: might attach link of the documentation for EN version instead of Chinese

File: docs/en/advanced_guides/datasets.md [hunk 6]
```diff
@@ -1 +1,143 @@
...
+│   │   │   │   ├── yyy{seg_map_suffix}
+│   │   │   ├── val
+│   │   │   │   ├── zzz{seg_map_suffix}
+
+```
+
+The image and ground truth annotation pair of `BaseSegDataset` should be of the same except suffix.
+A valid image and ground truth annotation filename pair should be like `xxx{img_suffix}` and `xxx{seg_map_suffix}` (extension is also included
+in the suffix). Two keys `data_list` and `metainfo` are introduced in `BaseSegDataset` to save raw data (such as image path)
+and meta data (such as class names of ground truth) information, respectively.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:37]: it would be better to replace these lines with an example to illustrate the item of `data_list` and metainfo

File: docs/en/advanced_guides/datasets.md [hunk 7]
```diff
@@ -1 +1,143 @@
...
+### Initialization
+
+Because of `BaseSegDataset` is inherited from [`BaseDataset`](https://github.com/open-mmlab/mmengine/blob/main/docs/zh_cn/advanced_tutorials/basedataset.md)
+in MMEngine, it also follows similar parts of initialization process of `BaseDataset`.
+
+Below is a table of some important methods in `BaseSegDataset` initialization method, which would be executed in order from top to bottom.
+
+| Methods executed in `BaseSegDataset` initialization |                            Usage                            |
+| :-------------------------------------------------: | :---------------------------------------------------------: |
+|                  `self._metainfo`                   |              Set meta information of dataset.               |
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:51]: `_metainfo` is an attribute not a function

File: docs/en/advanced_guides/datasets.md [hunk 8]
```diff
@@ -1 +1,143 @@
...
+|                `self._join_prefix()`                |            Join paths of images and annotations.            |
+|                   `self.pipeline`                   | Build data pipeline for data preprocessing and preparation. |
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:53]: pipeline is also an attribute

File: docs/en/advanced_guides/datasets.md [hunk 9]
```diff
@@ -1 +1,143 @@
...
+|          `self.filter_data()`          | Filter illegal data, such as data that has no annotations. |
+|   `self._get_unserialized_subset()`    |           Get subset data according to indices.            |
+|        `self._serialize_data()`        |   Serialize `data_list` of corresponding dataset class.    |
+
+In some cases such as visualization, only the meta information of the dataset is necessary, while loading annotation file
+is unnecessary. `BaseSegDataset` can skip loading annotations to save time by set `lazy_init=True`, in this case `self.full_init()` would not be executed.
+
+### Loading meta info
+
+Meta information is collected in `self._metainfo` method.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:71]: _metainfo is not a method

File: docs/en/advanced_guides/datasets.md [hunk 10]
```diff
@@ -1 +1,143 @@
...
+            except (TypeError, FileNotFoundError):
+                warnings.warn(f'{v} is not a meta file, simply parsed as '
+                              'meta information')
+                cls_metainfo[k] = v
+        else:
+            cls_metainfo[k] = v
+    return cls_metainfo
+```
+
+Then `_load_metainfo` returns parsed meta information for train/test process.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:94]: I think it is not necessary to provide the source code for load metainfo, and can be replaced with a table to list what will be loaded with different type of metainfo input

File: docs/en/advanced_guides/datasets.md [hunk 11]
```diff
@@ -1 +1,143 @@
...
+
+### Loading data info
+
+By default, `self.load_data_list()` would be called in `self.full_init()`. In `BaseSegDataset`,
+`self.load_data_list()` function is overwritten where annotation path `seg_map_path` would be added from annotation file directory or its meta file.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:99]: I think it would be better to add a table for 2 kinds of loading, and says '`load_data_list ` can be overwritten when it doesn't meet your datasets'.

File: docs/en/advanced_guides/datasets.md [hunk 12]
```diff
@@ -1 +1,143 @@
...
+
+Moreover, in `BaseSegDataset`, `label_map` and `reduce_zero_label` are also added in meta information dict:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:101]: this is not proper that explain label_map and reduce_zero_label in load data info section.

File: docs/en/advanced_guides/datasets.md [hunk 13]
```diff
@@ -1 +1,247 @@
-# Datasets
+# Dataset
+
+Due to the Dataset class in each OpenMMLab codebases have some common characteristics and requirements, such as efficient internal data storage format, support for the concatenation of different datasets, dataset repeated sampling,
+MMEngine implements `BaseDataset` which provides some basic interfaces and implements some DatasetWrappers with the same interfaces. The Dataset class of each OpenMMLab downstream codebase would inherit this base dataset class and add customization methods.
+
+The basic function of the MMEngine `BaseDataset` is loading the dataset information which has two categories.
+One is meta information, which represents the information related to the dataset itself and sometimes needs to be obtained by the model or other external components.
+The other is data information, which defines the file path and corresponding label information of specific data info.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:8]: it needs modification  `which defines the file path and corresponding label information of specific data info.`

File: docs/en/advanced_guides/datasets.md [hunk 14]
```diff
@@ -1 +1,247 @@
-# Datasets
+# Dataset
+
+Due to the Dataset class in each OpenMMLab codebases have some common characteristics and requirements, such as efficient internal data storage format, support for the concatenation of different datasets, dataset repeated sampling,
+MMEngine implements `BaseDataset` which provides some basic interfaces and implements some DatasetWrappers with the same interfaces. The Dataset class of each OpenMMLab downstream codebase would inherit this base dataset class and add customization methods.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:4]: I think it needs refinement

File: docs/en/advanced_guides/datasets.md [hunk 15]
```diff
@@ -1 +1,247 @@
...
+
+The basic function of the MMEngine `BaseDataset` is loading the dataset information which has two categories.
+One is meta information, which represents the information related to the dataset itself and sometimes needs to be obtained by the model or other external components.
+The other is data information, which defines the file path and corresponding label information of specific data info.
+
+More details of `BaseDataset` could be found in [MMEngine BaseDataset Documentation](https://github.com/open-mmlab/mmengine/blob/main/docs/en/advanced_tutorials/basedataset.md).
+
+## BaseSegDataset
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:12]: I think this subtitle is not necessary

File: docs/en/advanced_guides/datasets.md [hunk 16]
```diff
@@ -1 +1,247 @@
...
+
+In this document, we will introduce `BaseSegDataset` class in MMSegmentation.
+First, we would introduce standard dataset format in MMSegmentation, while creating a corresponding dataset class is also necessary.
+Then, we would introduce some important methods of `BaseSegDataset`: (1) initialization of `BaseSegDataset`, including detailed methods which would be called during this process.
+(2) loading meta information and (3) loading raw data information, the loading function of these two data information categories are basic function of `BaseSegDataset` for training and test process.
+Moreover, (4) `__getitem__` method used to give index of data information in `BaseSegDataset` initialization, which is similar to `torch.utils.data.Dataset`.
+
+More information about `BaseSegDataset` could be found in [basesegdataset.py](https://github.com/open-mmlab/mmsegmentation/blob/1.x/mmseg/datasets/basesegdataset.py).
+
+### Standard Dataset format
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:24]:
```suggestion
## Dataset Format
```

File: docs/en/advanced_guides/datasets.md [hunk 17]
```diff
@@ -1 +1,247 @@
...
+
+All datasets such as `CityscapesDataset` and `ADE20KDataset` are inherited from `BaseSegDataset` in the 1.x version of MMSegmentation.
+Some common methods are defined in `BaseSegDataset`,
+such as loading data information including meta information like categories and palette of dataset,
+and data information like path of image and ground truth, and modifying labels of dataset.
+
+`BaseSegDataset` is designed for supervised semantic segmentation task thus both images and annotations are necessary when model training.
+The directory structure of corresponding dataset is as below.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:32]: I think this section is about dataset format, so please move the 'Some common methods ....' to proper position

File: docs/en/advanced_guides/datasets.md [hunk 18]
```diff
@@ -1 +1,247 @@
...
+
+In OpenMMLab 2.0, MMEngine implements `BaseDataset` to provide some basic interfaces for downstream OpenMMLab codebase, which would inherit this base dataset class and add customization methods.
+
+The basic function of the MMEngine `BaseDataset` is loading the dataset information which has two categories.
+One is meta information, which represents the information related to the dataset itself and sometimes needs to be obtained by the model or other external components.
+The other is data information like the path of image and annotation files.
+
+More details of `BaseDataset` could be found in [MMEngine BaseDataset Documentation](https://github.com/open-mmlab/mmengine/blob/main/docs/en/advanced_tutorials/basedataset.md).
+
+`BaseSegDataset` is a base dataset class for semantic segmentation task in MMSegmentation v1.x, which is inherited from `BaseDataset` in MMEngine.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:11]:
```suggestion
`BaseSegDataset` is a base class of dataset for semantic segmentation task in MMSegmentation v1.x, which is inherited from `BaseDataset` in MMEngine.
```

File: docs/en/advanced_guides/datasets.md [hunk 19]
```diff
@@ -1 +1,247 @@
...
+#             'img_shape': (H, W),
+#             'metainfo': {
+#                     ...
+#                     }
+#             ...
+#         }
+# }
+```
+
+From above we can find a valid image and ground truth annotation filename pair should be like `xxx{img_suffix}`
```
Comment (MeowZheng) [inline docs/en/advanced_guides/datasets.md:87]:
```suggestion
A valid pair filenames of image and ground truth annotation should be like `xxx{img_suffix}`
```
